New features to patch, get_descriptor() and get_discriminated() methods for Struct and Union, new prophyc flag

Jiish opened the pull request:
Added possibility to patch union models into struct models. Also possible to change field and model names by "rename_field" and "rename_class" patches.

Added "--quiet" flag to prophyc compilation to suppress warning messages.

Commit:
Patch features:
-patch unions to structs
-change field names
-change model names

New --quiet flag to suppress warnings during compilation.

aurzenligl commented:
Hello, thank you for request!

I applied patch to master and committed on top of it (e.g. I allowed myself to change "rename_field/rename_class" name to "rename"). Please check these commits and see whether prophyc master still solves the problem correctly.

Do you want to have next prophy version released now or would you like to send more patches soon?

Jiish commented:
Alright thanks! I was thinking to do another patch which implements the class method to return Prophy message contents as a Python dictionary. I'll start working on it on Monday.

Commit:
Prophy messages now have fields() and values() built-in methods. fields() outputs message fields a a sorted list. values() outputs message contents as unsorted field-value Python dictionary.

Jiish commented:
Hi! I added methods fields() and values() to Struct and Union. The fields() method outputs the message fields in sorted order. values() outputs the message as a field-value Python dict.

aurzenligl commented on prophy/composite.py:
```diff
@@ -260,6 +260,44 @@ def field_to_string(name, type_, value):
     else:
```
"fields" and "values" do very specific thing, probably one party of prophy users will make use of it. My proposition is to extract "fields" and "values" outside of prophy to user's code and let this code work in a correct and stable way by establishing get_descriptor() function in union types and struct types api.
1. get_descriptor() function in struct/union generated classes
   this part is generic, will form a stable interface for users.
   who want reflection (iteration over fields, figuring out their type, etc.)
2. values() and fields() functions which are defined at user side (not in prophy),
   which may work in generic way with all structs and unions.

Doing 2 is quite straightforward, you can apply patch to fork master to see what I'm trying to say:
[values_fields_extraction.diff.txt](https://github.com/aurzenligl/prophy/files/315069/values_fields_extraction.diff.txt)
Since these functions can be defined anywhere, I propose to define them outside prophy. Using public/stable/official api (get_descriptors()) will allow any number of such functions to work correctly.

Now, let's try to figure out how this get_descriptors() could look like:
[get_descriptor.txt](https://github.com/aurzenligl/prophy/files/315081/get_descriptor.txt)

aurzenligl commented on prophy/composite.py:
```diff
@@ -295,6 +333,22 @@ def __str__(self):
                 out += field_to_string(name, tp, value)
         return out
 
+    def fields(self, prefix=''):
```
Fields or values may be valid message member names, which will clash with "fields" and "values" functions which were added.

aurzenligl commented on prophy/composite.py:
```diff
@@ -260,6 +260,44 @@ def field_to_string(name, type_, value):
     else:
         return "%s: %s\n" % (name, value)
```
Public API change should be reflected in prophy .rst documentation.

aurzenligl commented on prophy/composite.py:
```diff
@@ -260,6 +260,44 @@ def field_to_string(name, type_, value):
     else:
         return "%s: %s\n" % (name, value)
 
+def field_to_list(prefix, name, type_, value):
```
A lot of similar/identical parts of code in *list and *dict part.

aurzenligl commented on prophy/tests/test_field_value_output.py:
```diff
@@ -0,0 +1,133 @@
+def test_simple_union_fields(Union):
+    x = Union()
+    x.discriminator = 2
+    print(x)
```
Irrelevant print statement in test code.

Jiish commented:
Hi! Yes get_descriptor() is a good idea. I could use it then to implement external fields() and values() functions in the system where Prophy is used.

I also agree with your proposition of the get_descriptor() return value, which would include the type, constants and type of the class in question.

Commit:
New get_descriptor() and get_discriminated() API methods for prophy.Union and prophy.Struct.

get_descriptor() method returns an array of Descriptor objects for each field
the Struct contains. Each array member represents a field in the Struct.
A Descriptor object has name, type_ and kind attributes.
Kind is a simplified enumeration of the field type.
Kinds are defined as INT, ENUM, BYTES, ARRAY, STRUCT and UNION.

get_descriptor() can be called for Struct type objects, Struct instance objects and
Union type objects. get_discriminated() is called for Union instance object.
This method returns a Descriptor object for the currenly discriminated Union arm.

Jiish commented:
Here's a commit which implements get_descriptor() and get_discriminated() methods. I added a new module kind.py which contains enumerations for the field kinds. The kind class inherits Constant class which blocks assignments to the enumerations. I don't know if this was necessary but that was a way I thought could make the kind values seem like  "constants".

aurzenligl commented:
Thank you - I'll have it reviewed tomorrow.

aurzenligl commented on prophy/composite.py:
```diff
@@ -584,3 +594,12 @@ def __init__(cls, name, bases, attrs):
+    def __init__(self,name, type_, kind):
+            self.name = name
+            self.kind = kind
+            self.type_ = type_
```
Add a newline at end of file.

aurzenligl replied:
Consider changing self.type_ name to self.type.
PEP8 argues that names colliding with built-ins and keywords should be trailed with underscore, but in case of class member you don't get shadowing as you have to name enclosing object first:

```
type_  # this - as function argument - will shadow type builtin. Bad idea
self.type  # this doesn't shadow type built-in
field.type  # neither does this
```

aurzenligl commented on prophy/composite.py:
```diff
@@ -584,3 +594,12 @@ def __init__(cls, name, bases, attrs):
+
+class Descriptor(object):
+
+    def __init__(self,name, type_, kind):
```
Style self,name -> self, name

aurzenligl commented on prophy/composite.py:
```diff
@@ -584,3 +594,12 @@ def __init__(cls, name, bases, attrs):
             extend_union_descriptor(cls, descriptor)
             add_union_properties(cls, descriptor)
         super(union_generator, cls).__init__(name, bases, attrs)
+
```
Use only single blank line between two blocks of code (function and function, class and class, etc.).

aurzenligl commented on prophy/composite.py:
```diff
@@ -519,18 +529,18 @@ def __str__(self):
-        return out
+    def get_discriminated(self):
+        name, tp, _, _, _  =  self._discriminated
+        return Descriptor(name,tp,get_kind(tp))
```
In the sake of consistency with the rest of library, use "x, y" style instead of "x,y" when separating argument list.

aurzenligl commented on prophy/kind.py:
```diff
@@ -0,0 +1,16 @@
+class Constant(type):
```
Kind needn't have attribute setting blocked, it's a very unlikely error to happen,
and the method to disallow it is quite uncommon and may be surprising for reader.

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+
+def test_struct_instance_get_descriptor(DeeplyNestedStruct,NestedStruct,Struct):
+
+  fields = DeeplyNestedStruct().get_descriptor()
```
Please change code formatting to be consistent with this file and remaining files in this test suite:
4 spaces as an indent.

aurzenligl commented on prophy/composite.py:
```diff
@@ -584,3 +594,12 @@ def __init__(cls, name, bases, attrs):
+
+
+
+class Descriptor(object):
```
"Descriptor" is name used for class or union layout (I mean struct._descriptor field).
When we mean "the descriptor of single field", I think we should disambiguate and say: "FieldDesc" or "FieldDescriptor".

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+  for idx,f in enumerate(fields):
+    assert f.name   == ref[idx][0]
+    assert f.kind   == ref[idx][1]
+    assert f.type_  == ref[idx][2]
```
No newline.

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+
+  fields = DeeplyNestedStruct.get_descriptor()
+
+  ref = [['m', prophy.kind.kind.STRUCT, NestedStruct],
```
This should be "prophy.kind.STRUCT" instead of "prophy.kind.kind.STRUCT".
One "kind" needs to be removed. You can add "kind" to `__init__.py` file, so that users
only need to check for "prophy.kind.INT".

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+    assert "Cannot assign to a constant." == str(e.value)
+
+
+def test_struct_instance_get_descriptor(DeeplyNestedStruct,NestedStruct,Struct):
```
There is a lot of boilerplate, repetitive code in every testcase.
These tests may be one-liners:

```
def desc_to_tuples(desc):
    return [(fdesc.name, fdesc.kind, fdesc.type_) for fdesc in desc]

def test_struct_instance_get_descriptor(DeeplyNestedStruct, NestedStruct, Struct):
    assert desc_to_tuples(DeeplyNestedStruct().get_descriptor()) == [
        ('m', prophy.kind.kind.STRUCT, NestedStruct),
        ('n', prophy.kind.kind.STRUCT, Struct),
        ('o', prophy.kind.kind.INT, prophy.u32)
    ]
```

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+    assert f.kind == ref[idx][1]
+    assert f.type_ == ref[idx][2]
+
+def test_struct_type_get_descriptor(DeeplyNestedStruct,NestedStruct,Struct):
```
Style: "x,y" -> "x, y"

aurzenligl commented on prophy/composite.py:
```diff
@@ -584,3 +594,12 @@ def __init__(cls, name, bases, attrs):
+
+
+class Descriptor(object):
+
```
Consider adding `__repr__` method to make debugging simpler. Instead of:

```
(Pdb) Union.get_descriptor()[2]
<prophy.composite.Descriptor object at 0x7f1f78791c50>
```

Expression could return:

```
(Pdb) Union.get_descriptor()[2]
<a, INT>
```

You could also try to use namedtuple.
https://pymotw.com/2/collections/namedtuple.html

aurzenligl commented on prophy/tests/test_get_descriptor.py:
```diff
@@ -0,0 +1,107 @@
+  for idx,f in enumerate(fields):
+    assert f.name == ref[idx][0]
+    assert f.kind == ref[idx][1]
+    assert f.type_ == ref[idx][2]
```
Putting asserts in for loop makes it difficult to reason which loop iteration caused test to fail.
You can rewrite it to single assert comparing list of tuples with list of tuples.

aurzenligl commented on prophy/composite.py:
```diff
@@ -4,6 +4,7 @@
 from .exception import ProphyError
```
Please make sure that tox tests pass. Tox runs py.test tests in python 2.7 and 3.4. Currently the latter fails:
[tox.prophy.txt](https://github.com/aurzenligl/prophy/files/338197/tox.prophy.txt)
[tox.prophy_jiish.txt](https://github.com/aurzenligl/prophy/files/338216/tox.prophy_jiish.txt)

aurzenligl commented:
Hello, I just finished review:
- kind not added to prophy public API, which makes it difficult for user to check agains kinds,
- tests for python3 as run by tox are failing,
- documentation should be updated,
- a number of style (inconsistent formatting) / implementation (boilerplate) details.
  Please correct them and I'll pull the changeset.

Commit:
Issues fixed in previous commit

- code formatting fixed
- newlines added where they were missing
- tuples are now used in assertion checks in test_get_descriptor.py
- boilerplates removed in test_get_descriptor.py
- module "kind" imported to __init__.py
- Descriptor class renamed to FielDescriptor
- FieldDescriptor.kind is now a tuple containing a string and a value.
- Removed the obscure method to block kind class value manipulation
- FieldDescriptor __repr__ returns name, type and kind

Jiish commented:
Hi! Thank you for the comments! This commit should fix the aforementioned issues. The failing tests with Python3 were due the metaclass usage in kind.py.

I learnt that metaclass is defined as `class MyClass(metaclass=MyType)` in Python3 :+1: 

May I also write descriptions of the new features to the documentation?

aurzenligl commented:
Sorry for the delay, it's been a busy week.

I applied your commits onto master and released new version 0.7.7 - therefore I close the request.

Thanks :)
